packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts

```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates. All Rights Reserved.
+ */
+
+'use strict';
+
+import * as Types from "@hyperion/hyperion-util/src/Types";
+import performanceAbsoluteNow from "@hyperion/hyperion-util/src/performanceAbsoluteNow";
+import { ALChannelSurfaceMutationEvent, ALSurfaceMutationEventData } from "./ALSurfaceMutationPublisher";
+import * as ALSurfaceUtils from './ALSurfaceUtils';
+import { ALElementEvent, ALFlowletEvent, ALLoggableEvent, ALMetadataEvent, ALSharedInitOptions } from "./ALType";
+
+import * as ALEventIndex from './ALEventIndex';
+import { assert } from "@hyperion/hyperion-global";
+
+
+export type ALSurfaceVisibilityEventData =
+  ALFlowletEvent &
+  ALMetadataEvent &
+  ALElementEvent &
+  ALLoggableEvent &
+  Readonly<
+    {
+      surface: string;
+      event: 'component_visibility';
+      intersectionEntry: IntersectionObserverEntry;
+    }
+  >;
+
+export type ALChannelSurfaceVisibilityEvent = Readonly<{
+  al_surface_visibility_event: [ALSurfaceVisibilityEventData],
+}
+>;
+
+export type InitOptions = Types.Options<
+  ALSharedInitOptions<ALChannelSurfaceVisibilityEvent & ALChannelSurfaceMutationEvent>
+>;
+
+
+export function publish(options: InitOptions): void {
+  const { channel } = options;
+
+  // lookup surfaces that are mounted by their root element 
+  const activeSurfaces = new Map<Element | null, ALSurfaceMutationEventData>();
+
+  // We need one observer per threshold
+  const observers = new Map<number, IntersectionObserver>();
+
+  channel.addListener('al_surface_mutation_event', event => {
+    switch (event.event) {
+      case 'mount_component': {
+        if (event.capability?.trackVisibilityThreshold) {
+          const observer = getOrCreateObserver(event.capability.trackVisibilityThreshold);
+          const { element } = event;
+          /**
+           * IntersectionObserver cannot track display:content styles because
+           * these elements don't have their own "box".
+           * So, instead we have to focus on the children of the parent element
+           */
+          if (ALSurfaceUtils.isSurfaceWrapper(element)) {
+            for (let el = element.firstElementChild; el; el = el.nextElementSibling) {
+              observer.observe(el);
+            }
+          } else {
+            observer.observe(element);
+          }
+          activeSurfaces.set(element, event);
```

Comment:
Just trying to understand,

So here we set active surface to the parent element,  even if it's a wrapper that we are not observing?

Later this lookup,
```js
const surfaceEvent = activeSurfaces.get(element) ?? activeSurfaces.get(element.parentElement); // element or its parent, see above for .observe(...)
```
will correctly find the active surface, regardless of if it's a wrapper or one of the siblings of the parent?

Reply:
Yes, the problem is that we have to use `display:content` for the style of the wrapper to make sure it does not mess up the style of its children, in other words, it has no visual presentation; and exactly because of that its visibility cannot be detected!
So, when we have wrapper, we instead add the children of the wrapper, which almost always is actually just one child. 

Later, we know what the observer we have setup will only observe surface root nodes or direct descendants. Therefore, if we find either the observed node or its parent in the map, we can proceed